pip/src/noisy_simulator.rs

```diff
@@ -0,0 +1,375 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! This module is a thin `PyO3` wrapper around the rust `noisy_simulator` crate.
+
+use noisy_simulator::{ComplexVector, NoisySimulator, SquareMatrix};
+use num_complex::Complex;
+use pyo3::{exceptions::PyException, prelude::*};
+type PythonMatrix = Vec<Vec<Complex<f64>>>;
+
+pub(crate) fn register_noisy_simulator_submodule(py: Python, m: &PyModule) -> PyResult<()> {
+    // let m = PyModule::new(py, "noisy_simulator")?;
+    m.add("NoisySimulatorError", py.get_type::<NoisySimulatorError>())?;
+    m.add_class::<Operation>()?;
+    m.add_class::<Instrument>()?;
+    m.add_class::<DensityMatrixSimulator>()?;
+    m.add_class::<StateVectorSimulator>()?;
+    Ok(())
+}
+
+/// Performance Warning:
+///  nalgebra stores its matrices in column major order, and we want to send it
+///  to Python in row major order, this means that there will be lots of
+///  cache-misses in the convertion from one format to another.
+fn python_to_nalgebra_matrix(matrix: PythonMatrix) -> PyResult<SquareMatrix> {
+    let nrows = matrix.len();
+    let ncols = matrix[0].len();
+    // Check that matrix is well formed.
+    for row in &matrix {
+        if ncols != row.len() {
+            return Err(NoisySimulatorError::new_err(
+                "ill formed matrix, all rows should be the same length".to_string(),
+            ));
+        }
+    }
+    // Move matrix into a linear container.
+    let mut data = Vec::with_capacity(nrows * ncols);
+    for mut row in matrix {
+        data.append(&mut row);
+    }
+    Ok(SquareMatrix::from_row_iterator(nrows, ncols, data))
+}
+
+/// Performance Warning:
```

Comment:
Is there any guidance to go with this performance warning (here and above)? I see it used in one place, which seems unavoidable and not on a perf critical path. Is this here to stop folks accidentally using it later in other paths where it might be a concern?

Reply:
Yes, this is just used in a non-critical path for performance. But I wanted to avoid any perf accidents in the future.

Reply:
I will add that to the comment.